src/network/core/packet.h

```diff
@@ -38,49 +40,144 @@ typedef uint8  PacketType; ///< Identifier for the packet
 	NetworkSocketHandler *cs;
 
 public:
-	Packet(NetworkSocketHandler *cs);
+	Packet(NetworkSocketHandler *cs, size_t initial_read_size = sizeof(PacketSize));
 	Packet(PacketType type);
-	~Packet();
+
+	static void AddToQueue(Packet **queue, Packet *packet);
+	static Packet *PopFromQueue(Packet **queue);
 
 	/* Sending/writing of packets */
 	void PrepareToSend();
 
-	void Send_bool  (bool   data);
-	void Send_uint8 (uint8  data);
-	void Send_uint16(uint16 data);
-	void Send_uint32(uint32 data);
-	void Send_uint64(uint64 data);
-	void Send_string(const char *data);
+	bool   CanWriteToPacket(size_t bytes_to_write);
+	void   Send_bool  (bool   data);
+	void   Send_uint8 (uint8  data);
+	void   Send_uint16(uint16 data);
+	void   Send_uint32(uint32 data);
+	void   Send_uint64(uint64 data);
+	void   Send_string(const char *data);
+	size_t Send_bytes (const byte *begin, const byte *end);
 
 	/* Reading/receiving of packets */
-	void ReadRawPacketSize();
+	bool HasPacketSizeData() const;
+	bool ParsePacketSize();
+	size_t Size() const;
 	void PrepareToRead();
+	PacketType GetPacketType() const;
 
-	bool   CanReadFromPacket (uint bytes_to_read);
+	bool   CanReadFromPacket(size_t bytes_to_read, bool close_connection = false);
 	bool   Recv_bool  ();
 	uint8  Recv_uint8 ();
 	uint16 Recv_uint16();
 	uint32 Recv_uint32();
 	uint64 Recv_uint64();
 	void   Recv_string(char *buffer, size_t size, StringValidationSettings settings = SVS_REPLACE_WITH_QUESTION_MARK);
+
+	size_t RemainingBytesToTransfer() const;
+
+	/**
+	 * Transfer data from the packet to the given function. It starts reading at the
+	 * position the last transfer stopped.
+	 * See Packet::TransferIn for more information about transferring data to functions.
+	 * @param transfer_function The function to pass the buffer as second parameter and the
+	 *                          amount to write as third parameter. It returns the amount that
+	 *                          was written or -1 upon errors.
+	 * @param limit             The maximum amount of bytes to transfer.
+	 * @param destination       The first parameter of the transfer function.
+	 * @param args              The fourth and further parameters to the transfer function, if any.
+	 * @return The return value of the transfer_function.
+	 */
+	template <
+		typename A = size_t, ///< The type for the amount to be passed, so it can be cast to the right type.
+		typename F,          ///< The type of the function.
+		typename D,          ///< The type of the destination.
+		typename ... Args>   ///< The types of the remaining arguments to the function.
+	ssize_t TransferOutWithLimit(F transfer_function, size_t limit, D destination, Args&& ... args)
+	{
+		size_t amount = std::min(this->RemainingBytesToTransfer(), limit);
+		if (amount == 0) return 0;
+
+		/* Making buffer a char means casting a lot in the Recv/Send functions. */
+		const char *output_buffer = reinterpret_cast<const char*>(this->buffer.data() + this->pos);
```

Comment:
This still a bit icky to me, is reinterpret_cast really necessary? You've already got a `const byte*` from `.data()`

Reply:
If I change reinterpret_cast to static_cast:
src\network\core\packet.h(113): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
The problem is that byte = unsigned char, and char = signed char (at least for MSVC). Changing buffer to char breaks all the recv/send functions as there it's going to extend the sign bit or (signed) char so I would need to do a lot of casting there. TCP/UDP send/recv functions require (const) char *. So the alternative would be a C-style cast, but that doesn't seem the C++ way to me.